rust/hyperlane-base/src/types/multisig.rs

```diff
@@ -171,51 +203,218 @@ impl MultisigCheckpointSyncer {
+                                .and_then(|signer| {
+                                    weighted_validators
+                                        .iter()
+                                        .position(|vw| vw.validator == signer)
+                                })
+                                .map(|pos| (false, pos))
+                                .unwrap_or((true, 0)) // this shouldn't cause a panic because returning (true, 0) means that the validator is not in the weighted_validators list
+                        });
                         let checkpoint: MultisigSignedCheckpoint = signed_checkpoints.try_into()?;
                         debug!(checkpoint=?checkpoint, "Fetched multisig checkpoint");
                         return Ok(Some(checkpoint));
                     }
                 } else {
                     debug!(
-                        validator = format!("{:#x}", validator),
+                        validator = format!("{:#x}", vw.validator),
                         index = index,
                         "Unable to find signed checkpoint"
                     );
                 }
             } else {
-                debug!(%validator, "Unable to find checkpoint syncer");
+                debug!(%vw.validator, "Unable to find checkpoint syncer");
                 continue;
             }
         }
         debug!("No quorum checkpoint found for message");
         Ok(None)
     }
+
+    // this function fetches the highest index that meets the threshold weight
+    // it assumes sorted_indices is sorted in descending order of index
+    // it doesn't assume that the indices contain all validators in weighted_validators
+    fn fetch_highest_quorum_index(
+        &self,
+        weighted_validators: &[ValidatorWithWeight],
+        threshold_weight: Weight,
+        sorted_indices: &[(H256, u32)],
+    ) -> Option<u32> {
+        let weight_dict: HashMap<H256, Weight> = weighted_validators
+            .iter()
+            .map(|v| (v.validator, v.weight))
+            .collect();
+
+        let mut cumulative_weight: u128 = 0;
+        let mut validators_included = HashSet::new();
+
+        for (validator, index) in sorted_indices {
+            if !validators_included.contains(&validator) {
+                if let Some(weight) = weight_dict.get(validator) {
+                    cumulative_weight += weight;
+                    validators_included.insert(validator);
+                }
+            }
+
+            if cumulative_weight >= threshold_weight {
+                return Some(*index);
+            }
+        }
+
+        // If threshold is not met
+        debug!(
+            cumulative_weight = cumulative_weight,
+            threshold_weight = threshold_weight,
+            "Highest quorum index not found"
+        );
+        None
+    }
+
+    // signed_checkpoint.recover() is expensive, so we cache the result in recovered_addresses
+    fn fetch_or_recover(
```

Comment:
Hmm I think this function doesn't work for what we wanna do

The main problem is that the `let checkpoint_hash = signed_checkpoint.value.signing_hash();` is not unique for each validator. Like if we have 5 validators all behaving correctly, they'll each have signed the same hash. So going from signing hash -> validator address doesn't work - we'd end up putting the very first validator into the hashmap but not the other ones. This is why we have `let mut signed_checkpoints_per_root: HashMap<H256, Vec<SignedCheckpointWithMessageId>> = HashMap::new();`, where we can go from signing hash -> multiple validators